HPCC-17359 Roxie farmer ssl configuration

kenrowland opened the pull request:
Add necessary columns to ConfigMgr Roxie Ports tab to support
ssl Roxie farmer

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [x] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-17359 Roxie farmer ssl configuration

Add necessary columns to ConfigMgr Roxie Ports tab to support
ssl Roxie farmer

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

kenrowland commented:
@mckellyln Please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 83cde8c27defd801948031d864d525e83547c48c
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.796 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

mckellyln commented:
Is there any possible way to:
1). get the port number for a new port just added to be blank or something other than 9876 ?
2). have the protocol a pull-down menu of only 2 choices "native" or "ssl" ?

mckellyln commented:
after adding new Port, cursor changes to clock and never goes back to pointer.  All seems to work ok, but just wondering if this is a new issue ?

mckellyln commented:
If passphrase is empty string it is not written to xml.  Should it be written as passphrase="" ?

mckellyln commented:
Can we get the name=<> tag of each Farmer to get copied into the RoxieTopology.xml ?

mckellyln commented:
Should passphrase and protocol be a type <xs:string> instead of <xpathType> ?

mckellyln commented:
I accept this PR.  I just wish we could make the protocol a pull-down with the two options.
We should at the least have a note somewhere that "native" and "ssl" are the only accepted values.

richardkchapman commented:
Isn't that just a question of using:

              <xs:simpleType>
                <xs:restriction base="xs:string">
                  <xs:enumeration value="native"/>
                  <xs:enumeration value="ssl"/>
                </xs:restriction>
              </xs:simpleType>

and removing the type xs:string ? See other examples in roxie.xsd.in

richardkchapman commented:
@kenrowland Back to you

kenrowland commented:
The Ports section of the Roxie cluster is not fully controlled by the XSD configuration. It is generated by dynamically generated JavaScript code that is generated by C++ code that had to be modified in two places just to get the columns to appear. More research and time would be required to determine the format of the data necessary to make the entry box a pull down.

kenrowland commented:
Can we merge as is and open another issue to make this a pull down? Then as more is learned about the config manager we can look at making this change. Or, decide to wait for the replacement.

richardkchapman commented:
Can @miguelvazq or @GordonSmith help advise on the javascript?

GordonSmith commented:
js looks fine.

richardkchapman commented:
I was meaning, advise on how to make the js produce a droplist of the two options rather than a free-format text field

miguelvazq commented:
Since this is generated JavaScript I am assuming it is using vanilla JS. If it is YUI I feel I may be no help unfortunately as I am not familiar with it.

If vanilla you can do something of the sorts:

`var select = document.createElement('select');

data = [8080, 8010, 8005] // your data source

for (var i = 0; i < data.length; ++i) {  
  var option = new Option(data[i], data[i]);
  select.appendChild(option);
}

document.getElementById("port").appendChild(select);`

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 83cde8c27defd801948031d864d525e83547c48c
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.852 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
/opt/HPCCSystems/etc/init.d/hpcc_common: line 266: cfg.section.DEFAULT: command not found
/etc/init.d/hpcc-init: line 132: /sbin/hpcc_setenv: No such file or directory
Group for user  does not exist on the system. Exiting ...
HPCC status:
/opt/HPCCSystems/etc/init.d/hpcc_common: line 266: cfg.section.DEFAULT: command not found
/etc/init.d/hpcc-init: line 132: /sbin/hpcc_setenv: No such file or directory
Group for user  does not exist on the system. Exiting ...

Archive HPCC logs.
Removing RPM
Loaded plugins: fastestmirror
Resolving Dependencies
--> Running transaction check
---> Package hpccsystems-platform.x86_64 0:6.3.0-trunk0 will be erased
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package                   Arch        Version             Repository      Size
================================================================================
Removing:
 hpccsystems-platform      x86_64      6.3.0-trunk0        installed      168 M

Transaction Summary
================================================================================
Remove  1 Package

Installed size: 168 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
/opt/HPCCSystems/etc/init.d/hpcc_common: line 266: cfg.section.DEFAULT: command not found
/etc/init.d/hpcc-init: line 132: /sbin/hp
 ... ( comment is too long, 100 lines, 5132 bytes, truncated) ... 
                    

Complete!
Removing Directory - 
Removing Directory - 
Removing Directory - 
Removing Directory - 
Removing Directory - 
Removing Directory - 
Removing user - 
HPCC Uninstall: OK
Finish at: 2017-05-10_12-06-53
All done.

```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 83cde8c27defd801948031d864d525e83547c48c
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 95.466 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |621 |621 |0 |
|test (roxie) |748 |748 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on esp/files/scripts/configmgr/configmgr.js:
```diff
@@ -751,6 +751,18 @@ function initItemForRoxiePorts(item) {
+  item.passphrase_ctrlType = 0;
+  item.certificateFileName = "";
+  item.certificateFileName_extra = "";
+  item.certificateFileName_ctrlType = "";
```
this element seems to be populated a little differently than the others, was this intentionally done this way?

kenrowland replied:
Fixed

rpastrana commented on initfiles/componentfiles/configxml/roxie.xsd.in:
```diff
@@ -77,6 +77,42 @@
                 </xs:appinfo>
               </xs:annotation>
             </xs:attribute>
+             <xs:attribute name="protocol" type="xpathType" use="optional" default="native">
```
why are these attributes declared as type = xpathType?

rpastrana replied:
Also, are these XSD entries driving the Config mgr ui? if not, what's their purpose?

rpastrana commented on deployment/deployutils/deployutils.cpp:
```diff
@@ -513,6 +513,10 @@ class CGenerateJSFromXSD
         addItem(jsStrBuf, m_pEnv.get(), XML_TAG_ROXIE_FARM, TAG_REQARRAYTHREADS, "", 0, 1, "", 1);
         addItem(jsStrBuf, m_pEnv.get(), XML_TAG_ROXIE_FARM, "aclName", "", 0, 1, "|'#$process/ACL'", 4);
+        addItem(jsStrBuf, m_pEnv.get(), XML_TAG_ROXIE_FARM, "protocol", "", 0, 1, "", 1);
+        addItem(jsStrBuf, m_pEnv.get(), XML_TAG_ROXIE_FARM, "passphrase", "", 0, 1, "", 1);
```
Does it make sense to define protocol, passphrase, etc the same way TAG_PORT, TAG_REQARRAYTHREADS, etc are defined?

rpastrana replied:
Should the passphrases be masked?

kenrowland replied:
Successively tested different control type values until I found that a value of 5 represents a masked text entry box. Changed passphrase so it is now masked. Also documented other values.

kenrowland commented:
@rpastrana Please review latest changes. Note that XSD is not used for UI.

Commit:
HPCC-17359 Roxie farmer ssl configuration

Set passphrase to masked input text field.
Add drop menu for protocol.
Add name= to RoxieTopology farm process section

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

rpastrana commented on deployment/deployutils/deployutils.cpp:
```diff
@@ -332,6 +332,14 @@ void addItem(StringBuffer& jsStrBuf,
              const char* extra, 
              short ctrlType)
 {
+  // Control types:
```
Thanks for adding this comment!
It might be outside of the scope of this issue, but it'd be nice to have an enumeration to represent these types.

rpastrana commented on initfiles/componentfiles/configxml/roxie.xsd.in:
```diff
@@ -77,6 +77,41 @@
+                <xs:appinfo>
+                  <tooltip>Path to certificate filename</tooltip>
+                  <title>CertFile</title>
+                  <xpath>$process/certificateFileName</xpath>
```
let's clean up these attributes to be xs:strings. They prob don't need the XPATH element either.

rpastrana reviewed (commented):
@kenrowland a couple of minor comments.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 26f5022fa2cf00a89c519d58dc21a124a6c1de8d
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.473 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |622 |0 |
|test (roxie) |749 |749 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana reviewed (approved):
@kenrowland looks ok
